### PR TITLE
Multi model constraint manager

### PR DESCRIPTION
**Pull request by nv-braf**
Adding unit testing for constraint manager
Adding support for multi-model constraints
Adding additional checking of constraints in RCR test class

### CODE REVIEW

File: model_analyzer/result/constraint_manager.py [hunk 1]
```diff
@@ -62,10 +62,11 @@ def check_constraints(constraints, run_config_measurement):
         """
 
         if constraints:
-            for model_metrics in run_config_measurement.data():
+            for (i, model_metrics) in enumerate(run_config_measurement.data()):
                 for metric in model_metrics:
-                    if type(metric).tag in constraints:
-                        constraint = constraints[type(metric).tag]
+                    if constraints[i] is not None and type(
```
Comment (tgerdesnv) [inline model_analyzer/result/constraint_manager.py:67]: The function parameters list above states that constraints is a dict, but this seems to be a list?

File: model_analyzer/result/run_config_result.py [hunk 2]
```diff
@@ -43,15 +43,16 @@ def __init__(self, model_name, run_config, comparator, constraints=None):
             RunConfigResults and returns 1 if the first is better than
             the second, 0 if they are equal and -1
             otherwise
-        constraints: dict
+        constraints: dict (single model) or list of dicts (multi-model)
```
Comment (tgerdesnv) [inline model_analyzer/result/run_config_result.py:46]: We are probably relying on built-in types (dict/list) too much. I don't like having different types coming in to a function. Really seems like we need some kind of constraint class that hides a lot of implementation details. Likely not part of this story though.

File: tests/test_constraint_manager.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+            ConstraintManager.satisfies_constraints(constraints, rcm))
+
+        rcm = self._construct_rcm({"perf_throughput": 99})
+        self.assertFalse(
+            ConstraintManager.satisfies_constraints(constraints, rcm))
+
+    def test_multi_model_constraint_checks(self):
+        """
+        Test that satisfies_constraints works for a multi model
+        with a max style constraint
```
Comment (tgerdesnv) [inline tests/test_constraint_manager.py:172]: is this comment accurate? I see both min and max in the code

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: tests/test_constraint_manager.py (hunk 3 above)
Comment (nv-braf) [reply tests/test_constraint_manager.py:172]: No. Originally I was going to have separate tests, but decided it was easier to just have one test. Good catch.

File: model_analyzer/result/constraint_manager.py (hunk 1 above)
Comment (nv-braf) [reply model_analyzer/result/constraint_manager.py:67]: Updated parameters to indicate this is a list of dicts

File: model_analyzer/result/run_config_result.py (hunk 2 above)
Comment (nv-braf) [reply model_analyzer/result/run_config_result.py:46]: Yeah, I agree this is awkward

File: tests/test_constraint_manager.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+from model_analyzer.cli.cli import CLI
+
+from .common.test_utils import convert_to_bytes, construct_run_config_measurement
+
+import unittest
+from unittest.mock import MagicMock, patch
+from .common import test_result_collector as trc
+
+
+class TestConstraintManager(trc.TestResultCollector):
```
[the remark this answers is not in the document]
Comment (nv-braf) [reply tests/test_constraint_manager.py:28]: Thanks!

File: model_analyzer/result/run_config_result.py (hunk 2 above)
Comment (nv-braf) [reply model_analyzer/result/run_config_result.py:46]: TMA-732 has been filed to track this issue

**Review by tgerdesnv — APPROVED** [review]

**Pull request merged**